src/array.jl

```diff
@@ -943,16 +943,17 @@ function Base.sort!(v::CategoricalVector;
 
     # compute the order in which to read from counts
     ord = Base.Sort.ord(lt, by, rev, order)
-    index = eltype(v) >: Missing ? [missing; v.pool.valindex] : v.pool.valindex
     seen = counts .> 0
     anymissing = eltype(v) >: Missing && seen[1]
-    perm = sortperm(view(index, seen), order=ord)
-    nzcounts = counts[seen]
+    levs = eltype(v) >: Missing ? [missing; v.pool.valindex] : v.pool.valindex
+    sortedlevs = sort(Vector(view(levs, seen)), order=ord)
+    levelsmap = something.(indexin(sortedlevs, levs))
```

Comment:
why is this `something.` needed?

Reply:
That gets rid of the `Union{Nothing, Int}` eltype, so it should be better for performance in the loop, and it also ensures we get an error if something unexpected happens.

Reply:
Ah - I thought `indexin` does the narrowest possible type - indeed it always produces `Union{Nothing, Int}` so this is needed.